presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,160 @@
+
+    // It may take some time to remove a query's memory reservations from the worker nodes, that's why
+    // we check to see whether some time has passed after the query finishes to claim that it is leaked.
+    private static final int DEFAULT_LEAK_CLAIM_DELTA_SEC = 60;
+
+    // By default, leakedQueries will be a Set (created from a WeakHashMap -- just to be safe in case map gets large) of leaked query IDs.
+    @GuardedBy("this")
+    private final LeakedQueries leakedQueries;
+
+    // We don't rely on the size of the weak leakedQueries map for the number of leaked queries as we want an exact counter.
+    private final AtomicInteger numberOfLeakedQueries = new AtomicInteger();
+
+    static ClusterMemoryLeakDetector createMemoryLeakDetector()
+    {
+        return new ClusterMemoryLeakDetector(new LeakedQueries(true));
+    }
+
+    @VisibleForTesting
+    static ClusterMemoryLeakDetector createTestMemoryLeakDetector()
+    {
+        // For testing we want strong references, otherwise the tests will be non-deterministic due to GC effects.
+        return new ClusterMemoryLeakDetector(new LeakedQueries(false));
+    }
+
+    private ClusterMemoryLeakDetector(LeakedQueries leakedQueries)
+    {
+        this.leakedQueries = requireNonNull(leakedQueries, "leakedQueries is null");
+    }
+
+    /**
+     * @param queryExecutionsSupplier All queries that the coordinator knows about.
+     * @param queryMemoryReservations The memory reservations of queries in the GENERAL cluster memory pool.
+     */
+    void checkForMemoryLeaks(Supplier<List<QueryInfo>> queryInfoSupplier, Map<QueryId, Long> queryMemoryReservations)
+    {
+        requireNonNull(queryInfoSupplier);
+        requireNonNull(queryMemoryReservations);
+
+        Map<QueryId, Long> leakedQueryReservations = queryMemoryReservations.entrySet()
+                .stream()
+                .filter(Objects::nonNull)
+                .filter(entry -> entry.getValue() > 0)
+                .filter(entry -> isLeaked(queryInfoSupplier, entry.getKey()))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+
+        if (!leakedQueryReservations.isEmpty()) {
+            log.warn("Memory leak detected. The following queries are already finished, " +
+                    "but they have memory reservations on some worker node(s): %s", leakedQueryReservations);
+        }
+
+        // keep numberOfLeakedQueries and leakedQueries in sync
+        synchronized (this) {
+            numberOfLeakedQueries.set(leakedQueryReservations.size());
+            leakedQueries.update(leakedQueryReservations.keySet());
+        }
+    }
+
+    private static boolean isLeaked(Supplier<List<QueryInfo>> queryInfoSupplier, QueryId queryId)
+    {
+        boolean queryExists = false;
+        QueryState queryState = null;
+        DateTime queryEndTime = null;
+
+        for (QueryInfo queryInfo : queryInfoSupplier.get()) {
```

Comment:
This will fetch the list from the supplier for every query with a memory reservation, and then does a linear search.  How about using a supplier that does a lazy transform to a map to avoid the linear search, and does memoization to avoid the multiple fetches (and transforms)?